### ISSUE TITLE
CSRF vulnerability on OAuth approval page

### ISSUE THREAD

**Issue by jmandel** [grahamegrieve/fhirserver#17]
The "auth_choice" page can be invoked via HTTP GET as follows:

    https://argonaut.healthintersections.com.au/oauth2/auth_choice?form=true&userinfo=1&readClinical=1&readData=1&readMeds=1&readSchedule=1&writeSchedule=1&readAudit=1&readDocuments=1&readFinancial=1&readOther=1&writeOther=1

This means that my app can ask you to authorize access, and then repeatedly reload the approval URL above in a background tab; as soon as you sign in (and before you hit approve), the app will receive a fully-loaded authorization code.

The fix: CSRF token on the approval page.

**Comment by grahamegrieve** [comment]
Can you explain a little more about the fix please

**Comment by jmandel** [comment]
Of course! When you generate HTML for the form on `auth_choice`, you should include a hidden form parameter whose value is unguessable by an attacker (e.g. generate a random number, or generate a value tied to the user's session). Then, on submission of the approval form, ensure that the unguessable value is present alongside the guessable values like `form=true& writeOther=1`. So it'll be more like `form=true& writeOther=1&csrf-token=239879827981247`.

**Comment by grahamegrieve** [comment]
I'm not sure that I follow this. The client can only get the authorization code once the user commits the form - unless the client submits it for them. I don't see how a nonce changes this?

**Comment by jmandel** [comment]
> The client can only get the authorization code once the user commits the form - unless the client submits it for them

That's exactly right. And the "unless the client submits it for them" is the key distinction. If you provide a CSRF token (which internally you'd tie to the user's session cookie), this prevents the client from submitting it for them. Because the client can't know the CSRF token value.
